src/platform/Linux/MdnsImpl.cpp

```diff
@@ -167,6 +208,7 @@ void Poller::WatchFree(AvahiWatch * watch)
 
 void Poller::WatchFree(AvahiWatch & watch)
 {
+    (void) watch.mSocket.ReleaseFD();
```

Comment:
I'm a confused by this. WatchableSocket is normally expected to own the fd but not in the case of avahi?

If WatchableSocket is owning, why doesn't it close in its destructor?

If WatchableSocket is non-owning, why does it close the file descriptor in Close()?

If WatchableSocket non-owning but Close() is just a helper for release-and-close, should "release" call WatchableSocket::OnClose which does the actual detach-from-WatchableEventManager?

Reply:
> I'm a confused by this. WatchableSocket is normally expected to own the fd but not in the case of avahi?

Right. In the other cases (Inet and System::WakeEvent) the WatchableSocket effectively owns the fd, and replaces the fd from the point of view of the holder. In the Avahi case, Avahi itself owns the fd and lends it via `AvahiPoll.watch_new` and `watch_free`.

> If WatchableSocket is owning, why doesn't it close in its destructor?

For the normal (Inet) case, it's contained inside an IPEndPointBasis which is not constructible or destructible. It's contained in order to avoid requiring separate allocation. It may be worth revisiting that if/when the System/Inet virtualization and pool allocation improvements happen.

> If WatchableSocket is non-owning, why does it close the file descriptor in Close()?

Convenience for the common case.

> If WatchableSocket non-owning but Close() is just a helper for release-and-close, should "release" call WatchableSocket::OnClose which does the actual detach-from-WatchableEventManager?

Yes, `OnClose()` should be renamed `OnRelease()`.

Reply:
> > I'm a confused by this. WatchableSocket is normally expected to own the fd but not in the case of avahi?
> 
> Right. In the other cases (Inet and System::WakeEvent) the WatchableSocket effectively owns the fd, and replaces the fd from the point of view of the holder. In the Avahi case, Avahi itself owns the fd and lends it via `AvahiPoll.watch_new` and `watch_free`.

Having a non-owning watcher separate from fd-owner makes the most sense to me and would fit well with avahi but understood there are other constraints here.

> > If WatchableSocket is owning, why doesn't it close in its destructor?
> 
> For the normal (Inet) case, it's contained inside an IPEndPointBasis which is not constructible or destructible. It's contained in order to avoid requiring separate allocation. It may be worth revisiting that if/when the System/Inet virtualization and pool allocation improvements happen.

Ahh, the pool should really run destructors on slots released back to it :(

> 
> > If WatchableSocket is non-owning, why does it close the file descriptor in Close()?
> 
> Convenience for the common case.
> 
> > If WatchableSocket non-owning but Close() is just a helper for release-and-close, should "release" call WatchableSocket::OnClose which does the actual detach-from-WatchableEventManager?
> 
> Yes, `OnClose()` should be renamed `OnRelease()`.

Ah, I see that it does I was just confused by the name. Rename sounds good.

Reply:
> Having a non-owning watcher separate from fd-owner makes the most sense to me and would fit well with avahi but understood there are other constraints here.

I'll take a look at this when I do the remaining test and timer TODOs, which likely will be together with #7725. I don't know that it will make much difference; since we don't have weak fds, a watcher _must_ be informed before a file is closed.

> > Yes, `OnClose()` should be renamed `OnRelease()`.
> 
> Ah, I see that it does I was just confused by the name. Rename sounds good.

Ack, added to my TODO list.